### PR TITLE
Add linear API in test problems

### PR DESCRIPTION
**Pull request by tmigot**
#35

I kept the usual `jtprod` for problems with mixed constraints, so they have an implementation without allocation https://github.com/JuliaSmoothOptimizers/NLPModels.jl/issues/384

### CODE REVIEW

File: src/nlp/problems/hs14.jl [hunk 1]
```diff
@@ -136,6 +172,22 @@ function NLPModels.jtprod!(nlp::HS14, x::AbstractVector, v::AbstractVector, Jtv:
   return Jtv
 end
 
+function NLPModels.jtprod_lin!(nlp::HS14, x::AbstractVector, v::AbstractVector, Jtv::AbstractVector)
```
Comment (abelsiqueira) [inline src/nlp/problems/hs14.jl:175]: You forgot to remove `jtprod` above, so the lines below are missed.

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks for the PR. There are some coverage misses. At least one - I think more - is because you forgot to remove a function that obfuscates the `lin`-`nln` duality.
Also, I saw that CUTEst fails, but we haven't touched the API. Can you check why?

File: src/nlp/problems/hs14.jl (hunk 1 above)
Comment (tmigot) [reply src/nlp/problems/hs14.jl:175]: It was done on purpose, but you are right. I should probably remove that. Thanks

Comment (abelsiqueira) [reply src/nlp/problems/hs14.jl:175]: In that case you remove the `jprod` but not `jtprod`, so one of them would need to be changed.
To keep the general function and also the `lin` and `nln` versions, we need to implement the linear API tests and have them optional but off by default.

**Comment by tmigot** [comment]
The tests are failing because of the check on counters:
https://github.com/JuliaSmoothOptimizers/NLPModelsTest.jl/blob/e31cbdc5526e9d36ad870fc345b8bbf2f40b300c/src/nlp/consistency.jl#L76
I will update this PR.

**Comment by tmigot** [comment]
The `sum_counters` will break for problems that implement only `cons!` as its behavior changed https://github.com/JuliaSmoothOptimizers/NLPModels.jl/blob/94e14baa2cc8584eef2eeb6190b75fdecb7d4458/src/nlp/counters.jl#L77

**Review by abelsiqueira — APPROVED** [review]
Thanks for the update. LGTM

**Pull request merged**